File: node/tests/cross-lang-media.test.ts
/**
 * Cross-language media verify-parity tests (Node side).
 *
 * Mirrors `rust/haiai/tests/cross_lang_contract.rs`
 * (`cross_lang_signed_image_*` and `cross_lang_signed_text_md_*`) and
 * `python/tests/test_cross_lang_media.py`. Loads the same pre-signed
 * fixtures from `fixtures/media/signed.{png,jpg,webp,md}` (signed once by
 * the Rust regenerator in `rust/haiai/tests/regen_media_fixtures.rs`,
 * with its public key committed beside the media) and asserts
 * that the Node FFI verifyImage / verifyText paths return the same
 * Valid / HashMismatch verdicts.
 *
 * Any drift between languages here MUST be a parity bug, not a test-only
 * quirk — that is the entire point of this suite (PRD §5.5, TASK_011).
 *
 * Skips gracefully when the installed haiinpm native binding does not yet
 * expose the media methods (i.e., it predates TASK_008 napi-rs bindings —
 * rebuild via `cargo build -p haiinpm --release` to re-run).
 */

import { afterAll, beforeAll, describe, it, expect } from 'vitest';
import { createRequire } from 'node:module';
import { mkdtempSync, mkdirSync, readFileSync, readdirSync, realpathSync, statSync, copyFileSync, writeFileSync, rmSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';
import { createHash } from 'node:crypto';

const __filenameLocal = fileURLToPath(import.meta.url);
const __dirnameLocal = dirname(__filenameLocal);

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

const REPO_ROOT = resolve(__dirnameLocal, '../..');
const MEDIA_DIR = join(REPO_ROOT, 'fixtures', 'media');
const SIGNER_FIXTURE_PATH = join(MEDIA_DIR, 'SIGNER.json');
const CHECKSUMS_PATH = join(MEDIA_DIR, 'CHECKSUMS.txt');

// Password for the committed test-only verifier (not the media signer).
const VERIFIER_AGENT_PASSWORD = 'MediaFixtureVerifierPass!123';

// Set the password before any FFI module is loaded — JACS reads this at
// agent-construction time, and vitest's workers do not always propagate
// process.env mutations made later in the test lifecycle.
process.env.JACS_PRIVATE_KEY_PASSWORD = VERIFIER_AGENT_PASSWORD;

// ---------------------------------------------------------------------------
// FFI availability — skip when haiinpm pre-dates TASK_008.
// ---------------------------------------------------------------------------

interface NativeHaiClientCtor {
  new (configJson: string): {
    verifyImage?: (path: string, optsJson: string) => Promise<string>;
    verifyText?: (path: string, optsJson: string) => Promise<string>;
    signImage?: (inPath: string, outPath: string, optsJson: string) => Promise<string>;
    signText?: (path: string, optsJson: string) => Promise<string>;
    extractMediaSignature?: (path: string, optsJson: string) => Promise<string>;
  };
}

interface HaiinpmModule {
  HaiClient: NativeHaiClientCtor;
}

function loadHaiinpm(): HaiinpmModule | null {
  try {
    const dynamicRequire = createRequire(__filenameLocal);
    return dynamicRequire('haiinpm') as HaiinpmModule;
  } catch {
    return null;
  }
}

function probeMediaSupport(): boolean {
  const m = loadHaiinpm();
  if (!m) return false;
  try {
    const c = new m.HaiClient('{"jacs_id":"x"}');
    return (
      typeof c.verifyImage === 'function' &&
      typeof c.verifyText === 'function' &&
      typeof c.signImage === 'function' &&
      typeof c.signText === 'function' &&
      typeof c.extractMediaSignature === 'function'
    );
  } catch {
    return false;
  }
}

const MEDIA_SUPPORTED = probeMediaSupport();
const SKIP_REASON =
  'Installed haiinpm native binding does not expose the Layer-8 media methods. ' +
  'Rebuild via `cargo build -p haiinpm --release` then refresh ' +
  '`node/npm/haiinpm/haiinpm.darwin-arm64.node` once JACS 0.10.0+ is available.';

// ---------------------------------------------------------------------------
// Fixture-staging helpers — mirror Rust `cross_lang_contract.rs::stage_fixture_agent`.
// ---------------------------------------------------------------------------

interface SignerFixture {
  signer_id: string;
  algorithm: string;
  public_key_file: string;
  verifier_agent_dir: string;
}

function loadSigner(): SignerFixture {
  return JSON.parse(readFileSync(SIGNER_FIXTURE_PATH, 'utf-8')) as SignerFixture;
}

function loadChecksum(name: string): string {
  for (const line of readFileSync(CHECKSUMS_PATH, 'utf-8').split('\n')) {
    const parts = line.trim().split(/\s+/);
    if (parts.length === 2 && parts[1] === name) return parts[0];
  }
  throw new Error(`no checksum for ${name} in CHECKSUMS.txt`);
}

function readSignedWithChecksum(name: string): Buffer {
  const bytes = readFileSync(join(MEDIA_DIR, name));
  const expected = loadChecksum(name);
  const got = createHash('sha256').update(bytes).digest('hex');
  if (got !== expected) {
    throw new Error(`checksum drift on fixtures/media/${name}: got ${got}, expected ${expected}`);
  }
  return bytes;
}

function copyVerifierTree(src: string, dst: string, inAgentDir = false): void {
  mkdirSync(dst, { recursive: true });
  for (const name of readdirSync(src)) {
    const srcPath = join(src, name);
    const isDirectory = statSync(srcPath).isDirectory();
    // Git stores `{id}_{version}.json` because `:` is illegal on Windows.
    // Only restore agent-document filenames; preserve `public_keys`.
    const newName = !isDirectory && inAgentDir ? name.replace(/_/g, ':') : name;
    const dstPath = join(dst, newName);
    if (isDirectory) {
      copyVerifierTree(srcPath, dstPath, inAgentDir || name === 'agent');
    } else {
      copyFileSync(srcPath, dstPath);
    }
  }
}

interface StagedAgent {
  configPath: string;
  tmpDir: string;
  verificationKeyDir: string;
}

function stageVerifierAgent(signer: SignerFixture): StagedAgent {
  process.env.JACS_PRIVATE_KEY_PASSWORD = VERIFIER_AGENT_PASSWORD;

  const tmpDir = realpathSync(mkdtempSync(join(tmpdir(), 'haiai-media-parity-')));
  copyVerifierTree(join(REPO_ROOT, signer.verifier_agent_dir), tmpDir);
  const configPath = join(tmpDir, 'jacs.config.json');
  const verificationKeyDir = join(tmpDir, 'verification-keys');
  mkdirSync(verificationKeyDir);
  copyFileSync(
    join(REPO_ROOT, signer.public_key_file),
    join(verificationKeyDir, `${signer.signer_id}.public.pem`),
  );
  return { configPath, tmpDir, verificationKeyDir };
}

interface FFIClient {
  verifyImage: (path: string, optsJson: string) => Promise<string>;
  verifyText: (path: string, optsJson: string) => Promise<string>;
}

function buildFFIClient(signer: SignerFixture): {
  client: FFIClient;
  tmpDir: string;
  verifyOptionsJson: string;
} {
  const haiinpm = loadHaiinpm();
  if (!haiinpm) throw new Error('haiinpm not loadable');

  // The JACS native side reads `JACS_PRIVATE_KEY_PASSWORD` via Rust's
  // `std::env::var` (the OS environment block, not Node's `process.env`
  // proxy). vitest's default `threads` pool wraps env mutations so the
  // native side does not see them; this test runs in the `forks` pool via
  // `poolMatchGlobs` in `vitest.config.ts` so re-setting here is sufficient.
  process.env.JACS_PRIVATE_KEY_PASSWORD = VERIFIER_AGENT_PASSWORD;

  const staged = stageVerifierAgent(signer);
  const cfg = JSON.parse(readFileSync(staged.configPath, 'utf-8')) as Record<string, unknown>;
  const ffiConfig = JSON.stringify({
    jacs_id: (cfg.jacs_agent_id_and_version as string).split(':')[0],
    agent_name: 'MediaFixtureVerifier',
    agent_version: '1.0.0',
    key_dir: join(staged.tmpDir, cfg.jacs_key_directory as string),
    jacs_config_path: staged.configPath,
    base_url: 'http://localhost:1', // never used; verify is local-only
  });
  const native = new haiinpm.HaiClient(ffiConfig);
  return {
    client: native as unknown as FFIClient,
    tmpDir: staged.tmpDir,
    verifyOptionsJson: JSON.stringify({ key_dir: staged.verificationKeyDir }),
  };
}

// ---------------------------------------------------------------------------
// Tampering helpers — mirror Rust `tamper_after` + `tamper_text_body`.
// ---------------------------------------------------------------------------

function tamperAfter(buf: Buffer, marker: Buffer, offset: number): void {
  const idx = buf.indexOf(marker);
  if (idx === -1) throw new Error(`marker ${marker.toString('hex')} not found`);
  const target = idx + marker.length + offset;
  buf[target] ^= 0x01;
}

function tamperTextBody(buf: Buffer): void {
  const marker = Buffer.from('-----BEGIN JACS SIGNATURE-----', 'utf-8');
  const bodyEnd = buf.indexOf(marker);
  if (bodyEnd === -1) throw new Error('BEGIN marker present in signed.md');
  for (let i = bodyEnd - 1; i >= 0; i--) {
    const c = buf[i];
    if ((c >= 0x41 && c <= 0x5a) || (c >= 0x61 && c <= 0x7a)) {
      buf[i] ^= 0b0010_0000;
      return;
    }
  }
  throw new Error('no ASCII letter found before signature block');
}

// ---------------------------------------------------------------------------
// Per-suite state
// ---------------------------------------------------------------------------

let ffiClient: FFIClient | null = null;
let parityTmpDir: string | null = null;
let stageTmpDir: string | null = null;
let signer: SignerFixture | null = null;
let verifyOptionsJson = '{}';

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

const describeMaybe = MEDIA_SUPPORTED ? describe : describe.skip;

describeMaybe('cross-language media verify-parity (node)', () => {
  beforeAll(() => {
    if (!MEDIA_SUPPORTED) return;
    signer = loadSigner();
    const built = buildFFIClient(signer);
    ffiClient = built.client;
    parityTmpDir = built.tmpDir;
    verifyOptionsJson = built.verifyOptionsJson;
    stageTmpDir = realpathSync(mkdtempSync(join(tmpdir(), 'haiai-media-stage-')));
  });

  afterAll(() => {
    if (parityTmpDir) rmSync(parityTmpDir, { recursive: true, force: true });
    if (stageTmpDir) rmSync(stageTmpDir, { recursive: true, force: true });
  });

  if (!MEDIA_SUPPORTED) {
    // Single placeholder so the suite still appears in test output.
    it.skip(SKIP_REASON, () => {});
    return;
  }

  // -------------------------------------------------------------------------
  // verify_image — Node parity
  // -------------------------------------------------------------------------

  async function verifyImage(path: string): Promise<Record<string, unknown>> {
    return JSON.parse(await ffiClient!.verifyImage(path, verifyOptionsJson)) as Record<
      string,
      unknown
    >;
  }

  async function verifyText(path: string): Promise<Record<string, unknown>> {
    return JSON.parse(await ffiClient!.verifyText(path, verifyOptionsJson)) as Record<
      string,
      unknown
    >;
  }

  function stageSigned(name: string): string {
    const target = join(stageTmpDir!, name);
    writeFileSync(target, readSignedWithChecksum(name));
    return target;
  }

  function stageTampered(name: string, marker: Buffer, offset: number): string {
    const target = join(stageTmpDir!, name);
    const buf = Buffer.from(readSignedWithChecksum(name));
    tamperAfter(buf, marker, offset);
    writeFileSync(target, buf);
    return target;
  }

  it('signed.png verifies under fixture agent', async () => {
    const res = await verifyImage(stageSigned('signed.png'));
    expect(res.status).toBe('valid');
    expect(res.signer_id).toBe(signer!.signer_id);
  });

  it('signed.png tampered returns hash_mismatch', async () => {
    const path = stageTampered('signed.png', Buffer.from('IDAT', 'utf-8'), 6);
    const res = await verifyImage(path);
    expect(res.status).toBe('hash_mismatch');
  });

  it('signed.jpg verifies under fixture agent', async () => {
    const res = await verifyImage(stageSigned('signed.jpg'));
    expect(res.status).toBe('valid');
    expect(res.signer_id).toBe(signer!.signer_id);
  });

  it('signed.jpg tampered returns hash_mismatch', async () => {
    const path = stageTampered('signed.jpg', Buffer.from([0xff, 0xda]), 4);
    const res = await verifyImage(path);
    expect(res.status).toBe('hash_mismatch');
  });

  it('signed.webp verifies under fixture agent', async () => {
    const res = await verifyImage(stageSigned('signed.webp'));
    expect(res.status).toBe('valid');
    expect(res.signer_id).toBe(signer!.signer_id);
  });

  it('signed.webp tampered returns hash_mismatch', async () => {
    const path = stageTampered('signed.webp', Buffer.from('VP8L', 'utf-8'), 4);
    const res = await verifyImage(path);
    expect(res.status).toBe('hash_mismatch');
  });

  // -------------------------------------------------------------------------
  // verify_text — Node parity
  // -------------------------------------------------------------------------

  it('signed.md verifies under fixture agent', async () => {
    const target = join(stageTmpDir!, 'signed.md');
    writeFileSync(target, readSignedWithChecksum('signed.md'));
    const res = await verifyText(target);
    expect(res.status).toBe('signed');
    const sigs = (res.signatures as Array<Record<string, unknown>> | undefined) ?? [];
    expect(sigs).toHaveLength(1);
    expect(sigs[0].status).toBe('valid');
    expect(sigs[0].signer_id).toBe(signer!.signer_id);
  });

  it('signed.md tampered returns hash_mismatch', async () => {
    const target = join(stageTmpDir!, 'signed.md');
    const buf = Buffer.from(readSignedWithChecksum('signed.md'));
    tamperTextBody(buf);
    writeFileSync(target, buf);
    const res = await verifyText(target);
    expect(res.status).toBe('signed');
    const sigs = (res.signatures as Array<Record<string, unknown>> | undefined) ?? [];
    expect(sigs).toHaveLength(1);
    expect(sigs[0].status).toBe('hash_mismatch');
  });
});

File: go/cross_lang_media_test.go
//go:build cgo

// Cross-language media verify-parity tests (Go side).
//
// Mirrors `rust/haiai/tests/cross_lang_contract.rs`
// (`cross_lang_signed_image_*` and `cross_lang_signed_text_md_*`),
// `python/tests/test_cross_lang_media.py`, and
// `node/tests/cross-lang-media.test.ts`. Loads the same pre-signed fixtures
// from `fixtures/media/signed.{png,jpg,webp,md}` (signed once by the Rust
// regenerator in `rust/haiai/tests/regen_media_fixtures.rs`, with its public
// key committed beside the media) and asserts that the Go FFI
// VerifyImage / VerifyText paths return the same Valid / HashMismatch
// verdicts.
//
// Any drift between languages here MUST be a parity bug, not a test-only
// quirk — that is the entire point of this suite (PRD §5.5, TASK_011).
//
// CGo build only: needs `libhaiigo.dylib` (or .so) on the linker path.

package haiai

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

const verifierAgentPassword = "MediaFixtureVerifierPass!123"

func mediaRepoRoot(t *testing.T) string {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("Getwd: %v", err)
	}
	// go test runs in `go/`, repo root is one up.
	return filepath.Dir(wd)
}

func mediaDir(t *testing.T) string {
	return filepath.Join(mediaRepoRoot(t), "fixtures", "media")
}

// ---------------------------------------------------------------------------
// Signer fixture
// ---------------------------------------------------------------------------

type signerFixture struct {
	SignerID         string `json:"signer_id"`
	Algorithm        string `json:"algorithm"`
	PublicKeyFile    string `json:"public_key_file"`
	VerifierAgentDir string `json:"verifier_agent_dir"`
}

func loadSigner(t *testing.T) signerFixture {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(mediaDir(t), "SIGNER.json"))
	if err != nil {
		t.Fatalf("read SIGNER.json: %v", err)
	}
	var sf signerFixture
	if err := json.Unmarshal(raw, &sf); err != nil {
		t.Fatalf("decode SIGNER.json: %v", err)
	}
	return sf
}

// ---------------------------------------------------------------------------
// Checksum-verified fixture loader (mirrors Rust read_fixture_with_checksum)
// ---------------------------------------------------------------------------

func loadChecksum(t *testing.T, name string) string {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(mediaDir(t), "CHECKSUMS.txt"))
	if err != nil {
		t.Fatalf("read CHECKSUMS.txt: %v", err)
	}
	for _, line := range strings.Split(string(raw), "\n") {
		parts := strings.Fields(line)
		if len(parts) == 2 && parts[1] == name {
			return parts[0]
		}
	}
	t.Fatalf("no checksum for %s in CHECKSUMS.txt", name)
	return ""
}

func readSignedWithChecksum(t *testing.T, name string) []byte {
	t.Helper()
	bytes_, err := os.ReadFile(filepath.Join(mediaDir(t), name))
	if err != nil {
		t.Fatalf("read fixtures/media/%s: %v", name, err)
	}
	expected := loadChecksum(t, name)
	got := sha256.Sum256(bytes_)
	gotHex := hex.EncodeToString(got[:])
	if gotHex != expected {
		t.Fatalf("checksum drift on fixtures/media/%s: got %s, expected %s", name, gotHex, expected)
	}
	return bytes_
}

// ---------------------------------------------------------------------------
// Signed verifier staging — mirrors Rust `stage_media_verifier`.
// ---------------------------------------------------------------------------

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()
	_, err = io.Copy(out, in)
	return err
}

// copyVerifierTree preserves directory names such as `public_keys`, while
// restoring the `:` in agent document `{id}:{version}.json` filenames that
// Git stores with `_` for Windows compatibility.
func copyVerifierTree(src, dst string, inAgentDir bool) error {
	if err := os.MkdirAll(dst, 0o755); err != nil {
		return err
	}
	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		newName := entry.Name()
		if inAgentDir && !entry.IsDir() {
			newName = strings.ReplaceAll(newName, "_", ":")
		}
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, newName)
		if entry.IsDir() {
			if err := copyVerifierTree(srcPath, dstPath, inAgentDir || entry.Name() == "agent"); err != nil {
				return err
			}
		} else {
			if err := copyFile(srcPath, dstPath); err != nil {
				return err
			}
		}
	}
	return nil
}

type stagedAgent struct {
	configPath         string
	tmpDir             string
	verificationKeyDir string
}

func stageVerifierAgent(t *testing.T, signer signerFixture) stagedAgent {
	t.Helper()
	t.Setenv("JACS_PRIVATE_KEY_PASSWORD", verifierAgentPassword)

	tmpDir := t.TempDir()
	verifierDir := filepath.Join(mediaRepoRoot(t), signer.VerifierAgentDir)
	if err := copyVerifierTree(verifierDir, tmpDir, false); err != nil {
		t.Fatalf("stage signed verifier agent: %v", err)
	}
	configPath := filepath.Join(tmpDir, "jacs.config.json")
	verificationKeyDir := filepath.Join(tmpDir, "verification-keys")
	if err := os.MkdirAll(verificationKeyDir, 0o755); err != nil {
		t.Fatalf("create verification key directory: %v", err)
	}
	if err := copyFile(
		filepath.Join(mediaRepoRoot(t), signer.PublicKeyFile),
		filepath.Join(verificationKeyDir, signer.SignerID+".public.pem"),
	); err != nil {
		t.Fatalf("stage signer public key: %v", err)
	}

	return stagedAgent{
		configPath:         configPath,
		tmpDir:             tmpDir,
		verificationKeyDir: verificationKeyDir,
	}
}

// ---------------------------------------------------------------------------
// FFI client — wraps the staged fixture agent into a real `Client`.
// ---------------------------------------------------------------------------

func mediaParityVerifier(t *testing.T, signer signerFixture) (*Client, string) {
	t.Helper()

	staged := stageVerifierAgent(t, signer)
	cfgRaw, err := os.ReadFile(staged.configPath)
	if err != nil {
		t.Fatalf("re-read staged config: %v", err)
	}
	var cfg map[string]interface{}
	if err := json.Unmarshal(cfgRaw, &cfg); err != nil {
		t.Fatalf("decode staged config: %v", err)
	}
	idAndVer := cfg["jacs_agent_id_and_version"].(string)
	jacsID := strings.SplitN(idAndVer, ":", 2)[0]

	ffiConfig := map[string]interface{}{
		"jacs_id":          jacsID,
		"agent_name":       "MediaFixtureVerifier",
		"agent_version":    "1.0.0",
		"key_dir":          filepath.Join(staged.tmpDir, cfg["jacs_key_directory"].(string)),
		"jacs_config_path": staged.configPath,
		"base_url":         "http://localhost:1", // never used; verify is local-only
	}
	configJSON, err := json.Marshal(ffiConfig)
	if err != nil {
		t.Fatalf("marshal ffi config: %v", err)
	}
	ffi, err := newFFIClientFromConfig(string(configJSON))
	if err != nil {
		t.Skipf("Go FFI does not have the Layer-8 media methods loaded "+
			"(libhaiigo.dylib must be rebuilt against JACS 0.10.0+ source): %v", err)
	}
	cl, err := NewClient(WithFFIClient(ffi), WithJACSID(jacsID))
	if err != nil {
		t.Fatalf("new client: %v", err)
	}
	return cl, staged.verificationKeyDir
}

// mediaParityClient is shared by the Go sign-image/text tests. It uses the
// same current signed verifier fixture but does not need the external PQ key
// directory when signing and self-verifying newly created media.
func mediaParityClient(t *testing.T) *Client {
	t.Helper()
	client, _ := mediaParityVerifier(t, loadSigner(t))
	return client
}

// ---------------------------------------------------------------------------
// Tampering helpers — mirror Rust tamper_after / tamper_text_body.
// ---------------------------------------------------------------------------

func tamperAfter(t *testing.T, buf []byte, marker []byte, offset int) []byte {
	t.Helper()
	idx := bytes.Index(buf, marker)
	if idx == -1 {
		t.Fatalf("marker %x not found in fixture bytes", marker)
	}
	target := idx + len(marker) + offset
	out := append([]byte(nil), buf...)
	out[target] ^= 0x01
	return out
}

func tamperTextBody(t *testing.T, buf []byte) []byte {
	t.Helper()
	marker := []byte("-----BEGIN JACS SIGNATURE-----")
	bodyEnd := bytes.Index(buf, marker)
	if bodyEnd == -1 {
		t.Fatalf("BEGIN marker not present in signed.md")
	}
	out := append([]byte(nil), buf...)
	for i := bodyEnd - 1; i >= 0; i-- {
		c := out[i]
		if (c >= 0x41 && c <= 0x5A) || (c >= 0x61 && c <= 0x7A) {
			out[i] ^= 0x20
			return out
		}
	}
	t.Fatalf("no ASCII letter found before signature block")
	return nil
}

// ---------------------------------------------------------------------------
// Helpers — stage signed.* into a tempdir for verify
// ---------------------------------------------------------------------------

func writeStaged(t *testing.T, name string, data []byte) string {
	t.Helper()
	dir := t.TempDir()
	path := filepath.Join(dir, name)
	if err := os.WriteFile(path, data, 0o644); err != nil {
		t.Fatalf("write staged %s: %v", name, err)
	}
	return path
}

// ---------------------------------------------------------------------------
// Tests — verify_image / verify_text parity
// ---------------------------------------------------------------------------

func TestCrossLang_SignedImagePNGVerifies(t *testing.T) {
	signer := loadSigner(t)
	if signer.Algorithm != "pq2025" {
		t.Fatalf("fixture algorithm = %q, want pq2025 (parity baseline)", signer.Algorithm)
	}
	cl, keyDir := mediaParityVerifier(t, signer)

	path := writeStaged(t, "signed.png", readSignedWithChecksum(t, "signed.png"))
	got, err := cl.VerifyImage(context.Background(), path, VerifyImageOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if got.Status != MediaVerifyStatusValid {
		t.Fatalf("status = %q, want %q", got.Status, MediaVerifyStatusValid)
	}
	if got.SignerID == nil || *got.SignerID != signer.SignerID {
		t.Fatalf("signer_id = %v, want %q", got.SignerID, signer.SignerID)
	}
}

func TestCrossLang_SignedImagePNGTamperedReturnsHashMismatch(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	tampered := tamperAfter(t, readSignedWithChecksum(t, "signed.png"), []byte("IDAT"), 6)
	path := writeStaged(t, "signed.png", tampered)
	got, err := cl.VerifyImage(context.Background(), path, VerifyImageOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if got.Status != MediaVerifyStatusHashMismatch {
		t.Fatalf("status = %q, want %q (after tampering)", got.Status, MediaVerifyStatusHashMismatch)
	}
}

func TestCrossLang_SignedImageJPEGVerifies(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	path := writeStaged(t, "signed.jpg", readSignedWithChecksum(t, "signed.jpg"))
	got, err := cl.VerifyImage(context.Background(), path, VerifyImageOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if got.Status != MediaVerifyStatusValid {
		t.Fatalf("status = %q, want %q", got.Status, MediaVerifyStatusValid)
	}
	if got.SignerID == nil || *got.SignerID != signer.SignerID {
		t.Fatalf("signer_id = %v, want %q", got.SignerID, signer.SignerID)
	}
}

func TestCrossLang_SignedImageJPEGTamperedReturnsHashMismatch(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	tampered := tamperAfter(t, readSignedWithChecksum(t, "signed.jpg"), []byte{0xFF, 0xDA}, 4)
	path := writeStaged(t, "signed.jpg", tampered)
	got, err := cl.VerifyImage(context.Background(), path, VerifyImageOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if got.Status != MediaVerifyStatusHashMismatch {
		t.Fatalf("status = %q, want %q (after tampering)", got.Status, MediaVerifyStatusHashMismatch)
	}
}

func TestCrossLang_SignedImageWebPVerifies(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	path := writeStaged(t, "signed.webp", readSignedWithChecksum(t, "signed.webp"))
	got, err := cl.VerifyImage(context.Background(), path, VerifyImageOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if got.Status != MediaVerifyStatusValid {
		t.Fatalf("status = %q, want %q", got.Status, MediaVerifyStatusValid)
	}
	if got.SignerID == nil || *got.SignerID != signer.SignerID {
		t.Fatalf("signer_id = %v, want %q", got.SignerID, signer.SignerID)
	}
}

func TestCrossLang_SignedImageWebPTamperedReturnsHashMismatch(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	tampered := tamperAfter(t, readSignedWithChecksum(t, "signed.webp"), []byte("VP8L"), 4)
	path := writeStaged(t, "signed.webp", tampered)
	got, err := cl.VerifyImage(context.Background(), path, VerifyImageOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyImage: %v", err)
	}
	if got.Status != MediaVerifyStatusHashMismatch {
		t.Fatalf("status = %q, want %q (after tampering)", got.Status, MediaVerifyStatusHashMismatch)
	}
}

func TestCrossLang_SignedTextMDVerifies(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	path := writeStaged(t, "signed.md", readSignedWithChecksum(t, "signed.md"))
	got, err := cl.VerifyText(context.Background(), path, VerifyTextOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyText: %v", err)
	}
	if got.Status != VerifyTextStatusSigned {
		t.Fatalf("status = %q, want %q", got.Status, VerifyTextStatusSigned)
	}
	if len(got.Signatures) != 1 {
		t.Fatalf("len(signatures) = %d, want 1", len(got.Signatures))
	}
	sig := got.Signatures[0]
	if sig.Status != MediaVerifyStatusValid {
		t.Fatalf("sig.Status = %q, want %q", sig.Status, MediaVerifyStatusValid)
	}
	if sig.SignerID != signer.SignerID {
		t.Fatalf("sig.SignerID = %q, want %q", sig.SignerID, signer.SignerID)
	}
}

func TestCrossLang_SignedTextMDTamperedReturnsHashMismatch(t *testing.T) {
	signer := loadSigner(t)
	cl, keyDir := mediaParityVerifier(t, signer)

	tampered := tamperTextBody(t, readSignedWithChecksum(t, "signed.md"))
	path := writeStaged(t, "signed.md", tampered)
	got, err := cl.VerifyText(context.Background(), path, VerifyTextOptions{KeyDir: keyDir})
	if err != nil {
		t.Fatalf("VerifyText: %v", err)
	}
	if got.Status != VerifyTextStatusSigned {
		t.Fatalf("status = %q, want %q", got.Status, VerifyTextStatusSigned)
	}
	if len(got.Signatures) != 1 {
		t.Fatalf("len(signatures) = %d, want 1", len(got.Signatures))
	}
	sig := got.Signatures[0]
	if sig.Status != MediaVerifyStatusHashMismatch {
		t.Fatalf("sig.Status = %q, want %q (after tampering body)", sig.Status, MediaVerifyStatusHashMismatch)
	}
}

// Compile-time assertion the exported symbols we depend on are present in
// this build. If any of these go missing, this file will fail to compile,
// which is the right failure mode for a parity contract.
var _ = func() {
	var _ = MediaVerifyStatusValid
	var _ = MediaVerifyStatusHashMismatch
	var _ = VerifyTextStatusSigned
	var _ = VerifyImageOptions{}
	var _ = VerifyTextOptions{}
}
